src/shared/dissect-image.c

```diff
@@ -1409,56 +1414,101 @@ int decrypted_image_relinquish(DecryptedImage *d) {
+                        char *fn, *e, *n;
 
-                fn = newa(char, strlen(image) + STRLEN(".roothash") + 1);
-                n = stpcpy(fn, image);
-                e = endswith(fn, ".raw");
-                if (e)
-                        n = e;
+                        if (!IN_SET(r, -ENODATA, -EOPNOTSUPP, -ENOENT))
+                                return r;
 
-                strcpy(n, ".roothash");
+                        fn = newa(char, strlen(image) + STRLEN(".roothash") + 1);
+                        n = stpcpy(fn, image);
+                        e = endswith(fn, ".raw");
+                        if (e)
+                                n = e;
+                        else {
+                                e = endswith(fn, ".squashfs");
+                                if (e)
+                                        n = e;
+                        }
 
-                r = read_one_line_file(fn, &text);
-                if (r == -ENOENT) {
-                        *ret = NULL;
-                        *ret_size = 0;
-                        return 0;
+                        strcpy(n, ".roothash");
+
+                        r = read_one_line_file(fn, &text);
+                        if (r == -ENOENT) {
+                                *roothash = NULL;
+                                *roothash_size = 0;
+                                return 0;
+                        }
+                        if (r < 0)
+                                return r;
                 }
+
+                r = unhexmem(text, strlen(text), &k, &l);
                 if (r < 0)
                         return r;
+                if (l < sizeof(sd_id128_t))
+                        return -EINVAL;
+
+                *roothash = TAKE_PTR(k);
+                *roothash_size = l;
         }
 
-        r = unhexmem(text, strlen(text), &k, &l);
-        if (r < 0)
-                return r;
-        if (l < sizeof(sd_id128_t))
-                return -EINVAL;
+        if (hashdevice) {
+                r = getxattr_malloc(image, "user.verity.hashdevice", hashdevice, true);
+                if (r < 0) {
+                        char *e, *n;
+
+                        if (!IN_SET(r, -ENODATA, -EOPNOTSUPP, -ENOENT))
+                                return r;
+
+                        *hashdevice = (char *)malloc(strlen(image) + STRLEN(".hashdevice") + 1);
+                        if (!*hashdevice)
+                                return -ENOMEM;
+                        n = strcpy(*hashdevice, image);
+                        e = endswith(*hashdevice, ".raw");
+                        if (e)
+                                n = e;
+                        else {
+                                e = endswith(*hashdevice, ".squashfs");
+                                if (e)
+                                        n = e;
+                        }
+
+                        strcpy(n, ".hashdevice");
+
+                        r = access(*hashdevice, R_OK);
```

Comment:
hmm, is this check really desirable? if the file is missing we can detect that when we want to use, why do this ahead of time?

smells like toctou waiting to happen

Reply:
Just following the same pattern for the root hash, where the file is read immediately. Will remove it.